Add AWS warehouse with unit tests

mbarkley opened the pull request:
@cdvoisin This is the first PR of the first of two for the core AWS target adapter functionality.

[Here](https://docs.google.com/document/d/1T4wByP0Yp5HYQ9knE6161oiR4ZAtGUQsB9soVY2SeVs/edit#heading=h.uqsjrifbw50v) is a high-level overview of the functionality. Note that some things are not in this PR in an effort to reduce the size.


In this PR:
* AWS warehouse implementation that can mint tokens for S3 and Redshift
* Unit tests

Not in this PR:
* Key GC
* Background processes for key GC
* The Target Adapter interface implementation

chaopeng commented on lib/aws/warehouse.go:
```diff
@@ -0,0 +1,572 @@
+	})
+	if err != nil {
+		return fmt.Errorf("unable to create AWS user policy %s: %v", spec.principal.getId(), err)
+	} else {
```
no need else. same below

mbarkley commented:
@chaopeng Thanks for the feedback! I will take a first pass tomorrow.

mbarkley commented on lib/aws/client.go:
```diff
@@ -0,0 +1,91 @@
+	"github.com/aws/aws-sdk-go/service/sts"
+)
+
+type SdkApiClient struct {
```
I ended up making this type non-exported and renaming it to `sdkAPIClient` which is consistent with the guidelines [here](https://github.com/golang/go/wiki/CodeReviewComments#initialisms) regarding initialisms.

mbarkley commented on lib/aws/client.go:
```diff
@@ -0,0 +1,91 @@
+	"github.com/aws/aws-sdk-go/service/sts"
+)
+
+type SdkApiClient struct {
```
I ended up making this type not exported since it is only used inside the `aws` package, but I have added comments for all of the other exported types/functions that were suggested by golint.

mbarkley commented:
@chaopeng I've updated the PR based on your feedback. Additionally, I've applied your suggestions (with the help of golint) to other style mistakes you hadn't explicitly called out yet.

I've added these changes as a separate commit. Is that okay or do you prefer the changes to be squashed into a single commit?

chaopeng commented on lib/aws/warehouse_test.go:
```diff
@@ -0,0 +1,450 @@
+		return nil, err
+	}
+
+	var list []*iam.AccessKey
```
`list` defines but not use. Please delete or use it.

mbarkley replied:
@chaopeng Fixed.
